pkg/comp-functions/functions/helper/funcio.go

```diff
@@ -0,0 +1,236 @@
+			if err != nil {
+				return runtime.NewFatalErr(ctx, "cannot cleanup desired resources", err)
+			}
+		}
+
+		err = removeMetadata(&rawIO.Observed.Composite.Resource)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "cannote cleanup observed composite", err)
+		}
+
+		err = removeMetadata(&rawIO.Desired.Composite.Resource)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "cannot cleanup desired composite", err)
+		}
+
+		funcIO, err := json.Marshal(rawIO)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "cannot marshal functionIO to JSON", err)
+		}
+
+		currentIndex := meta.Generation % 2
+
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s-%d", meta.GetName(), funcName, currentIndex),
+				Namespace: "default",
+			},
+			Data: map[string]string{
+				"funcIO": string(funcIO),
+			},
+		}
+
+		err = diffFuncIOs(ctx, currentIndex, configMap, kubeClient, funcName)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "cannot diff funcIO", err)
+		}
+
+		err = createOrUpdateObject(ctx, configMap, kubeClient)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "can't save configmap containing functionIO", err)
+		}
+
+		return runtime.NewNormal()
+	}
+}
+
+func removeMetadata(res *k8sruntime.RawExtension) error {
+	obj := map[string]interface{}{}
+	err := json.Unmarshal(res.Raw, &obj)
+	if err != nil {
+		return err
+	}
+	if err != nil {
+		return err
+	}
+	// outer kubec objects
+	unstructured.RemoveNestedField(obj, "metadata", "generation")
+	unstructured.RemoveNestedField(obj, "metadata", "resourceVersion")
+	unstructured.RemoveNestedField(obj, "metadata", "annotations", "kubectl.kubernetes.io/last-applied-configuration")
+	unstructured.RemoveNestedField(obj, "metadata", "managedFields")
+
+	// inner observed objects
+	unstructured.RemoveNestedField(obj, "status", "atProvider", "manifest", "metadata", "generation")
+	unstructured.RemoveNestedField(obj, "status", "atProvider", "manifest", "metadata", "resourceVersion")
+	unstructured.RemoveNestedField(obj, "status", "atProvider", "manifest", "metadata", "annotations", "kubectl.kubernetes.io/last-applied-configuration")
+	unstructured.RemoveNestedField(obj, "status", "atProvider", "manifest", "metadata", "managedFields")
+
+	// inner spec objects
+	unstructured.RemoveNestedField(obj, "spec", "forProvider", "manifest", "metadata", "generation")
+	unstructured.RemoveNestedField(obj, "spec", "forProvider", "manifest", "metadata", "resourceVersion")
+	unstructured.RemoveNestedField(obj, "spec", "forProvider", "manifest", "metadata", "annotations", "kubectl.kubernetes.io/last-applied-configuration")
+	unstructured.RemoveNestedField(obj, "spec", "atProvider", "manifest", "metadata", "managedFields")
+
+	res.Raw, err = json.Marshal(&obj)
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func createOrUpdateObject(ctx context.Context, configMap *corev1.ConfigMap, kubeClient client.Client) error {
```

Comment:
Nitpick: Why don't you use https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

Reply:
Oh, I completely forgot, that this existed.